deltachat-ios/AppDelegate.swift

```diff
@@ -323,6 +333,34 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             }
         }
     }
+    
+    var shouldShutdownEventLoop: Bool = false
+    var eventShutdownSemaphore = DispatchSemaphore(value: 1)
+    func closeDB() {
+        // add flag to shutdown event handler
+        shouldShutdownEventLoop = true
+        // stopIo will generate atleast one event to the event handler can shut down
+        self.dcAccounts.stopIo()
+        // somehow wait for the eventhandler to be closed
+        eventShutdownSemaphore.wait()
+        // to avoid 0xdead10cc exceptions, we need to make sure there are no locks on files.
+        self.dcAccounts.closeDatabase()
+        shouldShutdownEventLoop = false
+    }
+    
+    func maybeStart() {
+        logger.debug("⏫ maybe openDatabase again 🤔")
+        if !dcAccounts.isOpen() {
```

Comment:
how can this ever happen? imu ...

- `isOpen()` checks if `accountsPointer != nil`

- on startup, `accountsPointer` is already set in `didFinishLaunchingWithOptions` (which should be always called before or instead any performFetch())

- on shutdown, `accountsPointer` is set to `nil` only when `AppDelegate`, where `maybeStart() `is part of, gets destroyed

have you ever seen this code being executed? otherwise, i suggest to revert the change as it add unneeded complexity

Reply:
> on shutdown, accountsPointer is set to nil only when AppDelegate, where maybeStart() is part of, gets destroyed

closeDB (which internally calls `self.dcAccounts.closeDatabase()`)  sets it to nil to shut it down and remove the locks.

Reply:
> have you ever seen this code being executed?

If you mean if I have seen its log messages? yes, quite often actually

Reply:
k, now i get the things slowly :) might have be easier to understand, with less code moved around, btw. so leaving continueDidFinishLaunchingWithOptions at its place